src/storage.zig

```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
+        }
+
+        fn spawn() std.Thread.SpawnError!void {
+            var threads = @maximum(1, std.Thread.getCpuCount() catch 1);
+            while (threads > 0) : (threads -= 1) {
+                const thread = try std.Thread.spawn(.{}, poll, .{&instance});
+                thread.detach();
+            }
+        }
+
+        fn submit(pool: *ThreadPool, next_tick: *NextTick) void {
+            pool.mutex.lock();
+            pool.queue.push(next_tick);
+            pool.unlock_after_queue_update();
+        }
+
+        fn poll(pool: *ThreadPool) void {
+            pool.mutex.lock();
+            while (true) {
+                while (pool.queue.pop()) |next_tick| {
+                    pool.unlock_after_queue_update();
+                    next_tick.callback(next_tick);
+                    pool.mutex.lock();
+                }
+
+                pool.idle += 1;
+                pool.cond.wait(&pool.mutex);
+                pool.idle -= 1;
+                pool.notified = false;
+            }
+        }
+
+        fn unlock_after_queue_update(pool: *ThreadPool) void {
+            if (!pool.queue.empty() and !pool.notified and pool.idle > 0) {
+                pool.notified = true;
+                pool.cond.signal();
+            }
+            pool.mutex.unlock();
+        }
+    };
+
     io: *IO,
     fd: os.fd_t,
+    main_thread_id: std.Thread.Id,
+
+    signal: Signal = undefined,
+    injected: Atomic(?*NextTick) = Atomic(?*NextTick).init(null),
 
-    next_tick_queue: FIFO(NextTick) = .{ .name = "storage_next_tick" },
-    next_tick_completion_scheduled: bool = false,
-    next_tick_completion: IO.Completion = undefined,
+    yield_queue: FIFO(NextTick) = .{ .name = "storage-yield" },
+    yield_completion: IO.Completion = undefined,
+    yield_scheduled: bool = false,
 
-    pub fn init(io: *IO, fd: os.fd_t) !Storage {
-        return Storage{
+    pub fn init(storage: *Storage, io: *IO, fd: os.fd_t) !void {
+        ThreadPool.init_once.call();
+        try ThreadPool.init_result;
+
+        try storage.signal.init(io, on_signal);
+        storage.* = .{
             .io = io,
             .fd = fd,
+            .main_thread_id = std.Thread.getCurrentId(),
+            .signal = storage.signal,
         };
     }
 
     pub fn deinit(storage: *Storage) void {
-        assert(storage.next_tick_queue.empty());
+        assert(storage.is_local());
+        storage.signal.deinit();
         assert(storage.fd != IO.INVALID_FILE);
         storage.fd = IO.INVALID_FILE;
     }
 
+    fn is_local(self: *const Storage) bool {
+        return std.Thread.getCurrentId() == self.main_thread_id;
+    }
+
     pub fn tick(storage: *Storage) void {
+        assert(storage.is_local());
         storage.io.tick() catch |err| {
             log.warn("tick: {}", .{err});
             std.debug.panic("io.tick(): {}", .{err});
         };
     }
 
+    pub const NextTickIntent = enum {
+        yield,
+        cpu_work,
```

Comment:
`work/inject` is not super obvious. Maybe `move_to_foreground/move_to_background`?

Reply:
Similarly `is_local` doesn't connect to `work/inject`. Maybe `is_foreground_thread`.

Reply:
Sure :)